### PR TITLE
added {get,set}_ctl_waitkill

### PR DESCRIPTION
**Pull request by mayank-mrinal-sony**

### CODE REVIEW

**Comment by ManaSugi** [comment]
Thanks! @mayank-mrinal-sony 
Could you check the MSRV test in CI?

**Comment by mayank-mrinal-sony** [comment]
> Thanks! @mayank-mrinal-sony Could you check the MSRV test in CI?

The MSRV CI test is failing due to [test.rs/306]( https://github.com/libseccomp-rs/libseccomp-rs/pull/237/files#diff-4bafee98e606237563dfe75c68dc30f5b3e9420f4eb918623340cdb95a54bea1R306). The env has api level set to 6 but 7 is required.
This test should have skipped because it's under `#[cfg(libseccomp_v2_6)]` but it didn't.

<details>
  <summary>'$ make test' result </summary>

```shell
$ make test
cargo test --all-features -- --color always --nocapture --test-threads 1
    Finished test [unoptimized + debuginfo] target(s) in 0.00s
     Running unittests src/lib.rs (target/debug/deps/libseccomp-19701cf213231586)

running 17 tests
test action::tests::test_parse_action ... ok
test api::tests::test_ensure_supported_api ... ok
test arch::tests::test_parse_arch ... ok
test arg_compare::tests::test_scmpargcompare ... ok
test compare_op::tests::test_parse_compare_op ... ok
test error::tests::test_debug ... ok
test error::tests::test_display ... ok
test error::tests::test_errno ... ok
test error::tests::test_from ... ok
test error::tests::test_msg ... ok
test error::tests::test_source ... ok
test error::tests::test_sysrawrc ... ok
test error::tests::test_with_msg ... ok
test error::tests::test_with_msg_and_source ... ok
test filter_attr::tests::test_parse_filter_attr ... ok
test filter_context::tests::test_as_ptr ... ok
test version::tests::test_ensure_supported_version ... ok

test result: ok. 17 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/global_reset.rs (target/debug/deps/global_reset-680f9789c4118485)

running 1 test
test test_reset_global_state ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/known_syscall_names.rs (target/debug/deps/known_syscall_names-2e1e74470aae7b71)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/notify.rs (target/debug/deps/notify-e979109017720bdb)

running 3 tests
test test_error ... ok
test test_resp_new ... ok
test test_user_notification ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.03s

     Running tests/tests.rs (target/debug/deps/tests-19e541b23a6d117f)

running 24 tests
test test_arch_functions ... ok
`test test_builder_pattern ... ok`
test test_check_api ... ok
test test_check_version ... ok
test test_display_syscall ... ok
test test_export_functions ... #
# pseudo filter code start
#
# filter for arch x86_64 (3221225534)
if ($arch == 3221225534)
  # default action
  action ALLOW;
# invalid architecture action
action KILL;
#
# pseudo filter code end
#
 >� 5@�����ok
test test_filter_attributes ... ok
test test_filter_reset ... ok
test test_get_api ... test_get_api: Got API level of 6
ok
test test_get_library_version ... test_get_library_version: 2.5.3
ok
test test_get_syscall_from_name ... ok
test test_get_syscall_name_from_arch ... ok
test test_merge_filters ... ok
test test_new_filter ... ok
test test_rule_add_array_load ... ok
test test_rule_add_exact_array_load ... ok
test test_rule_add_exact_load ... ok
test test_rule_add_load ... ok
test test_scmparch_native ... test_get_native_arch: native arch is X8664
ok
test test_set_api ... ok
test test_set_syscall_priority ... ok
test test_syscall_eq_i32 ... ok
test test_syscall_i32 ... ok
test test_syscall_new ... ok

test result: ok. 24 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/lib.rs (target/debug/deps/libseccomp_sys-3c28ebdc70cb1586)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests libseccomp

running 48 tests
test src/api.rs - api::check_api (line 84) ... ok
test src/api.rs - api::get_api (line 55) ... ok
test src/api.rs - api::set_api (line 33) ... ok
test src/arg_compare.rs - arg_compare::scmp_cmp (line 75) ... ok
test src/error.rs - error::SeccompError::sysrawrc (line 184) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::add_arch (line 167) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::add_rule (line 239) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::add_rule_conditional (line 277) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::add_rule_conditional_exact (line 366) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::add_rule_exact (line 328) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::export_bpf (line 1057) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::export_pfc (line 1029) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_act_badarch (line 536) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_act_default (line 510) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_api_sysrawrc (line 721) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_ctl_log (line 627) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_ctl_nnp (line 564) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_ctl_optimize (line 691) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_ctl_ssb (line 658) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_ctl_tsync (line 598) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::get_filter_attr (line 483) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::is_arch_present (line 133) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::load (line 409) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::merge (line 83) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::new (line 44) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::remove_arch (line 202) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::reset (line 1084) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_act_badarch (line 782) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_api_sysrawrc (line 1003) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_ctl_log (line 892) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_ctl_nnp (line 814) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_ctl_optimize (line 969) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_ctl_ssb (line 926) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_ctl_tsync (line 857) ... ok
test src/filter_context.rs - filter_context::ScmpFilterContext::set_syscall_priority (line 443) ... ok
test src/functions.rs - functions::reset_global_state (line 30) ... ok
test src/lib.rs - (line 18) ... ok
test src/lib.rs - (line 37) ... ok
test src/notify.rs - notify::ScmpNotifResp::new_continue (line 304) ... ignored
test src/notify.rs - notify::ScmpNotifResp::new_error (line 281) ... ignored
test src/notify.rs - notify::ScmpNotifResp::new_val (line 258) ... ignored
test src/syscall.rs - syscall::ScmpSyscall::from_name (line 123) ... ok
test src/syscall.rs - syscall::ScmpSyscall::from_name_by_arch (line 154) ... ok
test src/syscall.rs - syscall::ScmpSyscall::from_name_by_arch_rewrite (line 192) ... ok
test src/syscall.rs - syscall::ScmpSyscall::get_name (line 224) ... ok
test src/syscall.rs - syscall::ScmpSyscall::get_name_by_arch (line 251) ... ok
test src/syscall.rs - syscall::ScmpSyscall::new (line 84) ... ok
test src/version.rs - version::check_version (line 85) ... ok

test result: ok. 45 passed; 0 failed; 3 ignored; 0 measured; 0 filtered out; finished in 5.49s

   Doc-tests libseccomp-sys

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

```
</details>

**Comment by rusty-snake** [comment]
> This test should have skipped because it's under `#[cfg(libseccomp_v2_6)]`

Only if you use libseccomp `<=2.6.0`. We use libseccomp from it's main branch and force a version number of `9.9.9`.

<code>Error: Error { kind: Common("set_ctl_waitkill requires <b><i>libseccomp >= 2.6.0</i></b> and API level >= 7 (<b><i>current version: 9.9.9</i></b>, API level: 6)"), source: None, message: "set_ctl_waitkill requires libseccomp >= 2.6.0 and API level >= 7 (current version: 9.9.9, API level: 6)" }</code>

Actually the problem is that the CI runs with Linux 5.15 (`ubuntu-latest` = 22.04) but `SECCOMP_FILTER_FLAG_WAIT_KILLABLE_RECV` was added with Linux 5.19.

*And the musl test is broken on all PRs somehow.*

**Comment by rusty-snake** [comment]
> > This test should have skipped because it's under `#[cfg(libseccomp_v2_6)]`
> 
> Only if you use libseccomp `<=2.6.0`. We use libseccomp from it's main branch and force a version number of `9.9.9`.
> 
> `Error: Error { kind: Common("set_ctl_waitkill requires libseccomp >= 2.6.0 and API level >= 7 (current version: 9.9.9, API level: 6)"), source: None, message: "set_ctl_waitkill requires libseccomp >= 2.6.0 and API level >= 7 (current version: 9.9.9, API level: 6)" }`
> 
> Actually the problem is that the CI runs with Linux 5.15 (`ubuntu-latest` = 22.04) but `SECCOMP_FILTER_FLAG_WAIT_KILLABLE_RECV` was added with Linux 5.19.
> 
> _And the musl test is broken on all PRs somehow._

Maybe we could exclude them with a `cfg(github_actions)` but keep them enabled if run locally.

**Comment by ManaSugi** [comment]
@mayank-mrinal-sony Could you check @rusty-snake 's comments when you have time?

**Comment by mayank-mrinal-sony** [comment]
@ManaSugi @rusty-snake 

Made the changes ready for review.

**Comment by ManaSugi** [comment]
@mayank-mrinal-sony please check the conflict.

**Comment by mayank-mrinal-sony** [comment]
> @mayank-mrinal-sony please check the conflict.

Resolved

File: libseccomp/src/filter_context.rs [hunk 1]
```diff
@@ -1145,6 +1145,75 @@ impl ScmpFilterContext {
...
+    /// ```
+    /// # use libseccomp::*;
+    /// let mut ctx = ScmpFilterContext::new(ScmpAction::Allow)?;
+    /// # if check_api(7, ScmpVersion::from((2, 6, 0))).unwrap() {
+    /// ctx.set_ctl_waitkill(true)?;
+    /// assert!(!ctx.get_ctl_waitkill()?);
+    /// # }
+    /// # Ok::<(), Box<dyn std::error::Error>>(())
+    /// ```
+    pub fn get_ctl_waitkill(&self) -> Result<bool> {
```
[the remark this answers is not in the document]
Comment (rusty-snake) [reply libseccomp/src/filter_context.rs]: @ManaSugi WDYT?

File: libseccomp/tests/tests.rs [hunk 2]
```diff
@@ -287,6 +300,13 @@ fn test_builder_pattern() -> Result<(), Box<dyn std::error::Error>> {
         .set_api_sysrawrc(false)?
         .load()?;
 
+    #[cfg(github_actions)]
```
Comment (rusty-snake) [inline libseccomp/tests/tests.rs:303]: This way they are never executed. Can you add the following to `build.rs` and change the `cfg`s  to `#[cfg(not(github_actions))]`.

```rust
const GITHUB_ACTIONS: &str =  "GITHUB_ACTIONS";

fn main() {
    println!("cargo:rerun-if-env-changed={}", GITHUB_ACTIONS);

    [...]    

    if env::var(GITHUB_ACTIONS).as_deref() == Ok("true") {
        println!("cargo:rustc-cfg=github_actions");
    }
}
```

Comment (mayank-mrinal-sony) [reply libseccomp/tests/tests.rs]: I have made the changes and pushed it. 
@rusty-snake please review it.

**Comment by rusty-snake** [comment]
LGTM


Can you rebase to resolve the merge conflict.

**Comment by mayank-mrinal-sony** [comment]
> LGTM
> 
> Can you rebase to resolve the merge conflict.

@rusty-snake, I have resolved the conflice and pushed the commit.

**Review by rusty-snake — APPROVED** [review]

**Review by ManaSugi — APPROVED** [review]
@mayank-mrinal-sony LGTM, thank you!

File: libseccomp/src/filter_context.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ManaSugi) [reply libseccomp/src/filter_context.rs]: Yes, but `ensure_supported_api("get_ctl_waitkill", 7, ScmpVersion::from((2, 6, 0)))?;` prints the kind error message.

**Pull request merged**